ops/manifests/manifest.py

```diff
@@ -150,12 +154,46 @@ def default_release(self) -> str:
     @cached_property
     def latest_release(self) -> str:
         """Lookup the default release suggested by the manifest."""
-        return self.releases[0]
+        return self.releases[0] if self.releases else ""
 
     @property
     def current_release(self) -> str:
-        """Determine the current release from charm config."""
-        return self.config.get("release") or self.default_release or self.latest_release
+        """Determine the current release from charm config and available releases."""
+
+        # Determine what the user has configured (it could be undefined or empty)
+        config_release: str = self.config.get("release") or ""
+
+        if self._check_release:
+            if config_release:
+                if config_release not in self.releases:
+                    raise ManifestReleaseError(
+                        f"Configured release for {self.name} '{config_release}' not among {self.releases}"
+                    )
+                log.debug(
+                    f"Configured release for {self.name} '{config_release}' is among {self.releases}"
+                )
+            elif self.default_release:
+                if self.default_release not in self.releases:
+                    raise ManifestReleaseError(
+                        f"Default release for {self.name} '{self.default_release}' not among {self.releases}"
+                    )
+                log.debug(
+                    f"Default release for {self.name} '{self.default_release}' is among {self.releases}"
+                )
+            elif self.latest_release:
+                if self.latest_release not in self.releases:
+                    raise ManifestReleaseError(
+                        f"Latest release for {self.name} '{self.latest_release}' not among {self.releases}"
+                    )
+                log.debug(
+                    f"Latest release for {self.name} '{self.latest_release}' is among {self.releases}"
+                )
+            else:
+                raise ManifestReleaseError(
+                    f"No release selected for {self.name} among {self.releases}"
+                )
+
+        return config_release or self.default_release or self.latest_release
```

Comment:
Just a suggestion to reduce the nesting:

```suggestion
        if config_release:
            candidate = config_release
            source = "config"
        elif self.default_release:
            candidate = self.default_release
            source = "default"
        elif self.latest_release:
            candidate = self.latest_release
            source = "latest"
        else:
            raise ManifestReleaseError(f"No release selected for {self.name}, available: {self.releases}")
            
            
        if self._cehck_release and candidate not in self.releases:
            raise ManifestReleaseError(f"{source} release for {self.name} '{candidate}' is not among available releases: {self.releases}")
            
        log.debug("%s release for %s: '%s'. (Available: %s)", source, self.name, candidate, self.releases)

        return candidate
```

Reply:
thanks for this suggestion.  it was really clear once you shined a light on it